bus-mapping/src/circuit_input_builder.rs

```diff
@@ -0,0 +1,289 @@
+///    associated with each `eth_types::GethExecStep`] in the [`eth_types::GethExecTrace`].
+///
+/// The generated bus-mapping operations are:
+/// [`StackOp`](crate::operation::StackOp)s,
+/// [`MemoryOp`](crate::operation::MemoryOp)s and
+/// [`StorageOp`](crate::operation::StorageOp), which correspond to each
+/// [`OpcodeId`](crate::evm::OpcodeId)s used in each `ExecTrace` step so that the State Proof
+/// witnesses are already generated on a structured manner and ready to be added into the State
+/// circuit.
+pub struct CircuitInputBuilder {
+    pub block: Block,
+    pub block_ctx: BlockContext,
+}
+
+impl<'a> CircuitInputBuilder {
+    /// Create a new CircuitInputBuilder from the given `eth_block` and `constants`.
+    pub fn new<TX>(
+        eth_block: eth_types::Block<TX>,
+        constants: BlockConstants,
+    ) -> Self {
+        Self {
+            block: Block::new(&eth_block, constants),
+            block_ctx: BlockContext::new(),
+        }
+    }
+
+    /// Obtain a mutable reference to the state that the `CircuitInputBuilder` maintains,
+    /// contextualized to a particular transaction and a particular execution step in that
+    /// transaction.
+    pub fn state_ref(
+        &'a mut self,
+        tx: &'a mut Transaction,
+        tx_ctx: &'a mut TransactionContext,
+        step: &'a mut ExecStep,
+    ) -> CircuitInputStateRef {
+        CircuitInputStateRef {
+            block: &mut self.block,
+            block_ctx: &mut self.block_ctx,
+            tx,
+            tx_ctx,
+            step,
+        }
+    }
+
+    /// Handle a transaction with its corresponding execution trace to generate all the associated
+    /// operations.  Each operation is registered in `self.block.container`, and each step stores
+    /// the [`OperationRef`] to each of the generated operations.
+    pub fn handle_tx(
+        &mut self,
+        eth_tx: &eth_types::Transaction,
+        geth_trace: &GethExecTrace,
+    ) -> Result<(), Error> {
+        let mut tx = Transaction::new(eth_tx);
+        let mut tx_ctx = TransactionContext::new(eth_tx);
+        for (index, geth_step) in geth_trace.struct_logs.iter().enumerate() {
+            let mut step = ExecStep::new(geth_step, self.block_ctx.gc);
+            let mut state_ref = self.state_ref(&mut tx, &mut tx_ctx, &mut step);
+            geth_step.op.gen_associated_ops(
+                &mut state_ref,
+                &geth_trace.struct_logs[index..],
+            )?;
+            tx.steps.push(step);
+        }
+        self.block.txs.push(tx);
+        Ok(())
+    }
+}
+
+use crate::bytecode::Bytecode;
+use crate::evm::Gas;
+use crate::external_tracer;
+
+/// MockBlock is a type that contains all the information from a block required to build the
+/// circuit inputs.
+pub struct MockBlock {
+    pub eth_block: eth_types::Block<()>,
+    pub eth_tx: eth_types::Transaction,
+    pub block_ctants: BlockConstants,
+    pub geth_trace: eth_types::GethExecTrace,
+}
```

Comment:
If this is only going to be used for thesting, I would probably add it in a sepparated module under a #[cfg(test)] feature.

On that way it's less confusing.

Reply:
Thanks for the suggestion.  I moved all the mock generation functions and types to a new module called `mock`.  Not it looks more organized.  I haven't set that module under `#[cfg(test)]` because some of the mock functions are used by the `zkevm-circuits` crate!

Reply:
Seems much better! 

I'd file an issue and discuss in the next meeting or something similar what should we do with the `MockProver`, `MockTransaction` and and the testing facilities in general. :)